presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/SimplifyArrayOperationsRewriter.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.Session;
+import com.facebook.presto.SystemSessionProperties;
+import com.facebook.presto.sql.tree.Cast;
+import com.facebook.presto.sql.tree.ComparisonExpression;
+import com.facebook.presto.sql.tree.ComparisonExpression.Operator;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.ExpressionRewriter;
+import com.facebook.presto.sql.tree.ExpressionTreeRewriter;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.LambdaExpression;
+import com.facebook.presto.sql.tree.Literal;
+import com.facebook.presto.sql.tree.LongLiteral;
+import com.facebook.presto.sql.tree.NodeLocation;
+import com.facebook.presto.sql.tree.QualifiedName;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Optional;
+
+import static com.facebook.presto.sql.tree.ComparisonExpression.Operator.EQUAL;
+
+public class SimplifyArrayOperationsRewriter
+{
+    private SimplifyArrayOperationsRewriter() {}
+
+    public static Expression simplifyArrayOperations(Expression expression, Session session)
+    {
+        if (SystemSessionProperties.isSimplifyArrayOperations(session)) {
+            return ExpressionTreeRewriter.rewriteWith(new Visitor(), expression);
+        }
+
+        return expression;
+    }
+
+    private static class Visitor
+            extends ExpressionRewriter<Void>
+    {
+        private static boolean isConstant(Expression expression)
+        {
+            if (expression instanceof Cast) {
+                return isConstant(((Cast) expression).getExpression());
+            }
+
+            return expression instanceof Literal;
+        }
+
+        private static boolean isZero(Expression expression)
```

Comment:
cc @highker : Is this the recommended way to see whether a expression will be evaluated into certain constant?

Reply:
Not really; we need to use interpreter and evaluate the expression. The return result of the eval should be a const with `0` as its value.